File: mac/Voice/Hotkey/HotkeyInterpreter.swift
enum KeyEvent: Equatable {
    case flags(fn: Bool, rightOption: Bool, rightCommand: Bool)
    case keyDown(UInt16)
}

enum HotkeyAction: Equatable { case press, release, cancel }

/// Turns modifier/key events into press/release/cancel. Pure so the rules are testable.
struct HotkeyInterpreter {
    static let escapeKeyCode: UInt16 = 53
    let choice: HotkeyChoice
    // D2: exposed so HotkeyMonitor can tell whether a `setChoice` while this key is held must be
    // deferred instead of orphaning the in-flight press.
    private(set) var isHeld = false

    /// True for the duration of a hands-free session (prd-hands-free-dictation.md rule 6).
    ///
    /// During a hold, any other key can only be a shortcut — Fn+←, a chord — so cancelling is
    /// right. Latched mode is the opposite situation: the whole point is that the user keeps
    /// typing while it records, so the session must survive every keystroke. Esc is the single
    /// exception, because "throw this away" has no other expression once the key is no longer held.
    var latched = false

    init(choice: HotkeyChoice) { self.choice = choice }

    mutating func handle(_ e: KeyEvent) -> HotkeyAction? {
        switch e {
        case .flags(let fn, let ro, let rc):
            let down: Bool
            switch choice { case .fn: down = fn; case .rightOption: down = ro; case .rightCommand: down = rc }
            if down && !isHeld { isHeld = true; return .press }
            if !down && isHeld { isHeld = false; return .release }
            return nil
        case .keyDown(let code):
            // Rule 6: a latched session ignores the keyboard entirely except for Esc. Note the key
            // is *not* held during a latched session, so the `isHeld` guard below would already
            // swallow every keyDown — this branch exists to let Esc back through, not to block the
            // others.
            if latched { return code == Self.escapeKeyCode ? .cancel : nil }
            // Any real key while the hotkey is held means a shortcut (Fn+F5, Fn+←, Esc): not a
            // dictation. H1: `isHeld` resets right here, not at the swallowed physical release —
            // `.cancel` is the last action this press produces; the flags(down: false) that follows
            // it must fall through to `return nil` below (isHeld is already false by then), which is
            // also why a separate `cancelled` flag is no longer needed.
            guard isHeld else { return nil }
            isHeld = false
            return .cancel
        }
    }
}

File: windows/Spit.Core/Hotkey/HotkeyInterpreter.cs
namespace Spit.Core;

// Port of mac/Voice/Hotkey/HotkeyInterpreter.swift. Keys are adapted to Windows (rule 27): the Mac's
// `flags(fn:rightOption:rightCommand:)` becomes the state of the two Windows hotkeys, and key codes
// are Windows virtual-key codes. `HotkeyTranslator` produces these from raw hook events.

public abstract record KeyEvent
{
    private KeyEvent() { }
    /// Modifier state of the two right-hand hotkeys, like the Mac's `flagsChanged`.
    public sealed record Flags(bool RightCtrl, bool RightAlt) : KeyEvent;
    /// A non-modifier key went down. `VkCode` is a Windows virtual-key code.
    public sealed record KeyDown(int VkCode) : KeyEvent;
}

public enum HotkeyAction { Press, Release, Cancel }

/// Turns modifier/key events into press/release/cancel. Pure so the rules are testable.
public sealed class HotkeyInterpreter
{
    /// VK_ESCAPE; the Mac's `escapeKeyCode` is 53, its kVK_Escape.
    public const int EscapeKeyCode = 0x1B;

    public HotkeyInterpreter(HotkeyChoice choice) => Choice = choice;

    public HotkeyChoice Choice { get; }

    /// Exposed so the monitor can tell whether a hotkey change while this key is held must be deferred
    /// instead of orphaning the in-flight press.
    public bool IsHeld { get; private set; }

    /// True for the duration of a hands-free session (prd-hands-free-dictation.md rule 6).
    ///
    /// During a hold, any other key can only be a shortcut, so cancelling is right. Latched mode is the
    /// opposite: the user keeps typing while it records, so the session must survive every keystroke.
    /// Esc is the single exception, because "throw this away" has no other expression once the key is
    /// no longer held.
    public bool Latched { get; set; }

    public HotkeyAction? Handle(KeyEvent e)
    {
        switch (e)
        {
            case KeyEvent.Flags f:
                var down = Choice == HotkeyChoice.RightAlt ? f.RightAlt : f.RightCtrl;
                if (down && !IsHeld) { IsHeld = true; return HotkeyAction.Press; }
                if (!down && IsHeld) { IsHeld = false; return HotkeyAction.Release; }
                return null;

            case KeyEvent.KeyDown k:
                // A latched session ignores the keyboard except for Esc. The key is not held during a
                // latched session, so the `IsHeld` guard below would already swallow every key-down —
                // this branch exists to let Esc back through.
                if (Latched) return k.VkCode == EscapeKeyCode ? HotkeyAction.Cancel : null;
                // Any real key while the hotkey is held means a shortcut: not a dictation. `IsHeld`
                // resets here, not at the physical release — `Cancel` is the last action this press
                // produces, and the Flags(down: false) that follows falls through to `return null`.
                if (!IsHeld) return null;
                IsHeld = false;
                return HotkeyAction.Cancel;

            default:
                return null;
        }
    }
}
